PERF: delay preset sync until tab completion / attribute access, track file desync

tangkong opened the pull request:
## Description
Delays preset file reading until tab-completion has been attempted.  Stashes this information on the device instance itself.  This is not the default behavior, and you must opt into this by supplying the optional argument `defer_loading` to `setup_presets_paths`

Also tracks the modification time of the preset file on sync, making the `Preset` instance the source of truth to see if the presets need to be synced.

## Motivation and Context
Presets have taken a very long time to load recently, and the first place to look is often file i/o.  This delays this as long as possible while attempting to preserve the behavior hutch scientists use: tab completion.

This has also been made optional.  Both `setup_presets_paths` and `Presets.sync()` now take an optional `defer_loading` argument, that if true will skip the preset data loading step

In a nutshell:
* the device's Preset instance tracks whether the preset is out of sync with the files.  This is done by tracking file modification times

* The device checks if a sync is necessary, and if so syncs with the files when:
  * `__dir__` is called (during tab-completion)
  * `__getattribute__` is called with a preset-related prefix (`wm_`, `umv_`, `mv_`)

`Presets.sync()` is also called on device init, since it's part of `Presets.__init__`.  This happens both on device-creation and `setup_presets_path`, though on device-creation we don't do any file loading.  This is a ~0.013s double-hit, split between the device load and presets load sections for `FltMvInterface` devices

## How Has This Been Tested?
Tests pass

Tested interactively through hutch-python, loading some subset of 559 xcs presets
* this PR: 7.53s (0.01345s / preset)
* current master: 16.21s (0.0289s / preset)

There's some variation but I'm not going to do a battery of tests here

### Why isn't this just 0 s?
There's still some path access going on in hutch-python, it's just stopping short of actually opening the presets file.

## Where Has This Been Documented?
This PR, a stray comment

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on GitHub Actions
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

tangkong commented:
I did some more digging (manual profiling because I'm dumb) as to what causes the remaining add-methods to take so long, and I think it boils down to `stat()` calls having awful performance.  I'll do some more digging but I do think this is a contribution to be considered separately

ZLLentz commented on pcdsdevices/interface.py:
```diff
@@ -899,25 +904,29 @@ def set_position(self, position):
 
 
-def setup_preset_paths(**paths):
+def setup_preset_paths(**kwargs):
```
Nitpick: marginally simpler
```suggestion
def setup_preset_paths(defer_loading: bool = False, **paths):
```

ZLLentz commented on pcdsdevices/interface.py:
```diff
@@ -1140,6 +1152,7 @@ def _register_method(self, obj, method_name, method):
         self._methods.append((obj, method_name))
         setattr(obj, method_name, MethodType(method, obj))
         if hasattr(obj, '_tab'):
+            # obj._tab: TabCompletionHelperInstance
```
Nitpick: lingering commented code line

tangkong replied:
I kinda thought that would be useful for future us as a sort of pseudo type-hint

ZLLentz commented on pcdsdevices/interface.py:
```diff
@@ -899,25 +904,29 @@ def set_position(self, position):
         Presets._paths[k] = Path(v)
     for preset in Presets._registry:
-        preset.sync()
+        preset.sync(defer_loading=defer_loading)
```
Suggestion: maybe it's simpler to just not call sync at all if `defer_loading` is set to `True`, as long as this doesn't cause any unexpected issues. Then, `sync` doesn't need any code changes.

tangkong replied:
The other half of `sync` sets up the `add_*` methods, which I thought would be nice to have in all cases.  But if we're ok removing those methods then I could get behind it.

ZLLentz replied:
I forgot about that. I'm not immediately sure how I feel about those methods in the context of having not loaded the presets at all yet.

ZLLentz commented on pcdsdevices/interface.py:
```diff
@@ -227,8 +227,13 @@ def __init_subclass__(cls, **kwargs):
     def __dir__(self):
+        if not self._tab_initialized and hasattr(self, 'presets'):
+            self._tab_initialized = True
+            self.presets.sync()
```
Change/addition request: we need to identify other times where presets may need a late sync. For example, if someone is expecting presets to be available in a hutch-python function call they need a clear way to trigger this without tabbing (or we need to load them automatically).

tangkong replied:
The `device.presets.sync()` call is still available, and seems to be how we suggest people sync presets across sections.  I imagine we haven't advertised this as thoroughly as necessary.

tangkong replied:
I guess we could overwrite `__getattribute__` and catch attempts to access `mv_*`, `umv_*`, and `wm_*` methods, but are there any other access modes we can think of?

ZLLentz replied:
Is it possible that doing this on `__getattribute__` is actually most correct to make sure the loaded values are always up-to-date?

tangkong replied:
My one qualm with this is that it does break the already fragile composition abstraction barrier we've set up with the Presets module.  Maybe there's also a way around this.  

Back to the mines!

ZLLentz commented on pcdsdevices/interface.py:
Side-side thought: maybe this is an opportunity to look at https://jira.slac.stanford.edu/browse/ECS-923. (In short, since presets are only loaded once, they desync quickly between different sessions)

tangkong replied:
Indeed, I think now is probably the time.  A quick investigation shows that file modification queries are indeed faster than the entire sync (perhaps obvious now, but it answers the question in ECS-923)

```python
In [10]: %%time
    ...: print(list(os.path.getmtime(p) for p in xcs_ccm.x.presets._paths.values()))
    ...: 
    ...: 
[1736549297.5101147, 1736549297.5232627]
CPU times: user 0 ns, sys: 399 µs, total: 399 µs
Wall time: 2.19 ms

In [11]: %%time
    ...: xcs_ccm.x.presets.sync()
    ...: 
    ...: 
CPU times: user 15.7 ms, sys: 1.31 ms, total: 17 ms
Wall time: 65.7 ms
```

It should be simple enough to stash these values and re-sync if they fall out of date

tangkong replied:
I've given this a shot.  It was a bit squirrely to test, but I think I ironed it out

ZLLentz reviewed (commented):
I like this and have a lot of opinions about presets in general.


> stat() calls having awful performance

I agree this is a problem on NFS/WEKA.
Probably the main way for us to mitigate this would be reworking presets to use fewer files, but that has its own annoyances.

tangkong commented:
Updated with some rather significant changes, now we store the last file modification times with every `sync()` call, and use that to decide whether or not to update.  Also added some tests to codify this

ZLLentz commented:
I think at least one of your commits has not been pushed

tangkong commented:
I blame github

tangkong commented:
Fingers crossed that we don't see race conditions like in #1050 , and the resulting #1055

ZLLentz commented on pcdsdevices/tests/test_interface.py:
```diff
@@ -191,6 +201,55 @@ def test_presets_type(presets, fast_motor):
+    # deferred_fast_motor_preset must come last,
+    # to clear cache after motor is created (and sync-ed at init)
+    assert fast_motor.presets.sync_needed
+    fast_motor.__dir__()  # mimic tab completion request
```
Nitpick: is `dir(fast_motor)` equivalent?

tangkong replied:
I believe so, do we prefer that to the dunder-access?  I used `__dir__` just to match the code usage.

ZLLentz reviewed (approved):
I like how this is implemented and I like how it speeds up the startup preset loading by 2x

janeliu-slac commented:
First time seeing updates to pcdsdevices, so probably don't have good feedback this time. Are presets for device settings? Why does preset file reading need to be delayed until tab completion has been attempted?

tangkong commented:
Presets are basically position shortcuts that users can save/load.  They're dynamically added to the devices attributes, and are backed by simple yaml files.
https://pcdshub.github.io/pcdsdevices/v8.7.0/presets.html

Because each device is backed by a file, we end up opening and reading a lot of files in some hutch-python sessions.  So instead of loading them all at once, we defer loading until the user actually needs it

ZLLentz commented on pcdsdevices/interface.py:
```diff
@@ -1304,6 +1328,17 @@ def state(self):
         return state
 
+    @property
+    def sync_needed(self) -> bool:
```
One last tiny tiny nitpick: this is a property that touches the filesystem, usually I prefer things that may take time to be function calls. Usually we expect attribute access to be fast, so if they are not fast we can accidentally write code with performance issues.

tangkong replied:
A good point.  I'll do that

ZLLentz reviewed (approved):
I'm excited for the speedup
